File: euclidean_primer/euclidean_primer__fast_exponentiation.py
# 実行時間 10ms
a, b, m = list(map(int, input().split()))
o = 1

while 0 < b:
    if (b & 1) == 1:
        o = o * a % m
    a = a * a % m
    b >>= 1
print(o)

File: euclidean_primer/euclidean_primer__fast_exponentiation.cs
using static System.Console;

internal class Program
{
    static void Main()
    {
        ulong[] abm = System.Array.ConvertAll( ReadLine().Split(), ulong.Parse );
        ulong o = 1;

        while( 0 < abm[1] ) {
            if( ( abm[1] & 1 ) == 1 ) {
                o = o * abm[0] % abm[2];
            }
            abm[0] = abm[0] * abm[0] % abm[2];
            abm[1] >>= 1;
        }
        WriteLine( o );
        return;
    }
}
